supabase/functions/mailgun/index.ts

```diff
@@ -0,0 +1,184 @@
+        contact = existingContact;
+    } else {
+        // If the contact does not exist, we need to create it, along with the company if needed
+        const { data: existingCompany, error: fetchCompanyError } =
+            await supabaseAdmin
+                .from('companies')
+                .select('*')
+                .eq('name', domain)
+                .maybeSingle();
+        if (fetchCompanyError)
+            return new Response(
+                `Could not fetch companies from database, name: ${domain}`,
+                { status: 500 }
+            );
+
+        // deno-lint-ignore no-explicit-any
+        let company: any = undefined;
+        if (existingCompany) {
+            company = existingCompany;
+        } else {
+            const { data: newCompanies, error: createCompanyError } =
+                await supabaseAdmin
+                    .from('companies')
+                    .insert({ name: domain, sales_id: sales.id })
+                    .select();
+            if (createCompanyError)
+                return new Response(
+                    `Could not create company in database, name: ${domain}`,
+                    { status: 500 }
+                );
+            company = newCompanies[0];
+        }
+
+        // Create the contact
+        const { data: newContacts, error: createContactError } =
+            await supabaseAdmin
+                .from('contacts')
+                .insert({
+                    first_name: firstName,
+                    last_name: lastName,
+                    email,
+                    company_id: company.id,
+                    sales_id: sales.id,
+                    first_seen: new Date(),
+                    last_seen: new Date(),
+                    tags: [],
+                })
+                .select();
+        if (createContactError || !newContacts[0])
+            return new Response(
+                `Could not create contact in database, email: ${email}`,
+                { status: 500 }
+            );
+        contact = newContacts[0];
+    }
+
+    // Add note to new contact
+    const { error: createNoteError } = await supabaseAdmin
+        .from('contactNotes')
+        .insert({
+            contact_id: contact.id,
+            text: noteContent,
+            sales_id: sales.id,
+        });
+    if (createNoteError)
+        return new Response(
+            `Could not add note to contact ${email}, sales ${salesEmail}`,
+            { status: 500 }
+        );
+
+    return new Response('OK');
+});
+
+/**
+ * Extracts the first name, last name, email, and domain from a mail contact (e.g. "To" or "From").
+ *
+ * Example:
+ *   From: "Alice Cooper <alice@sandbox.mailgun.org>"
+ *
+ * Return Value:
+ *  {
+ *    firstName: "Alice",
+ *    lastName: "Cooper",
+ *    email: "alice@sandbox.mailgun.org",
+ *    domain: "sandbox.mailgun.org"
+ * }
+ *
+ * @param to
+ * @returns
+ */
+const extractMailContactData = (to: string) => {
```

Comment:
this one deserves unit tests with partial emails

Reply:
It's planned!